### PR TITLE
Issue1 lucas ch

### PR DESCRIPTION
**Pull request by lucas-ch**
create pull request

### CODE REVIEW

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+import org.eclipse.emf.common.util.EList;
+
+import fr.adaussy.permadeler.model.Permadeler.Area;
+import fr.adaussy.permadeler.model.Permadeler.CompatibilityLink;
+import fr.adaussy.permadeler.model.Permadeler.Genus;
+import fr.adaussy.permadeler.model.Permadeler.Root;
+import fr.adaussy.permadeler.model.utils.EMFUtils;
+
+public class ValidationService {
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java:27]: Please add a java doc

Something like. Class gathering validation diagram on reprensentation

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+	public static boolean validatePlantationCompatibility(Area self) {
+
+		// get the list of incompatible species
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java]: Please remove commented code

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+import fr.adaussy.permadeler.model.Permadeler.Area;
+import fr.adaussy.permadeler.model.Permadeler.CompatibilityLink;
+import fr.adaussy.permadeler.model.Permadeler.Genus;
+import fr.adaussy.permadeler.model.Permadeler.Root;
+import fr.adaussy.permadeler.model.utils.EMFUtils;
+
+public class ValidationService {
+
+	public static boolean validatePlantationCompatibility(Area self) {
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java:29]: Please add java doc. Didn't you get a warning from checkstyle?

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+		Set<Genus> plantedGenus = self.getPlantations().stream()//
+				.filter(e -> e.getType() != null && e.getType().getType() != null)//
+				.map(e -> e.getType().getType())//
+				.collect(Collectors.toSet());
+
+		return !root.getSeedLib().getCompatibilityMatrix().getCompatibilties().stream().anyMatch(l -> {
+			return isIncompatibleLink(plantedGenus, l);
+		});
+		// Set<CompatibilityLink> compatibilityLinkNeg = new HashSet<>();
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java]: Please remove commented code

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import fr.adaussy.permadeler.model.Permadeler.Root;
+import fr.adaussy.permadeler.model.utils.EMFUtils;
+
+public class ValidationService {
+
+	public static boolean validatePlantationCompatibility(Area self) {
+
+		// get the list of incompatible species
+
+		// session = SessionManager.INSTANCE.getSession(self).getSemanticCrossReferencer()
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java]: Please refactor using ECrossReferencer

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+import fr.adaussy.permadeler.model.Permadeler.Species;
+import fr.adaussy.permadeler.model.Permadeler.Zone;
+
+public class ValidationServiceTest {
+
+	private static final PermadelerFactory sFactory = PermadelerFactory.eINSTANCE;
+
+	private static final int nbPlantations = 3;
+
+	private Root root = sFactory.createRoot();
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java:30]: On a test class that will test several service please don't use attributes of the test class. Instead you may create a InnerClass that will hold the code required for your test.

**Comment by lucas-ch** [comment]
dcsqfdsq

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+import fr.adaussy.permadeler.model.Permadeler.Genus;
+import fr.adaussy.permadeler.model.Permadeler.KnowledgeBase;
+import fr.adaussy.permadeler.model.Permadeler.PermadelerFactory;
+import fr.adaussy.permadeler.model.Permadeler.Plantation;
+import fr.adaussy.permadeler.model.Permadeler.Planting;
+import fr.adaussy.permadeler.model.Permadeler.Root;
+import fr.adaussy.permadeler.model.Permadeler.Species;
+import fr.adaussy.permadeler.model.Permadeler.Zone;
+
+public class ValidationServiceTest {
```
[the remark this answers is not in the document]
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java:37]: done

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java (hunk 1 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java:27]: done

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java (hunk 2 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java]: done

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java (hunk 3 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java:29]: done

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java (hunk 4 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java]: done

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java (hunk 5 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java]: I dont kno hox to do this

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java (hunk 6 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java]: done

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+		TestServices.plantPlantations(TestServices.plantations);
+
+		boolean validation = ValidationService.validatePlantationCompatibility(TestServices.area);
+		assertTrue(validation == true);
+	}
+
+	/**
+	 * Variables and methods used in ValidationServiceTest
+	 */
+	private static class TestServices {
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java:130]: Please rename to "InputModelHelper"

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+		private final static PermadelerFactory sFactory = PermadelerFactory.eINSTANCE;
+
+		private static final int nbPlantations = 3;
+
+		private static Root root = sFactory.createRoot();
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java:135]: All attributes should not be static.
You should instantiate in each test a "InputModelHelper inputModel = new InputModelHelper()"

Then either access attribute using get or direct attribute

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Arthur Daussy.
+ *
+ * This program and the accompanying materials are made 
+ * available under the terms of the Eclipse Public License 2.0 
+ * which is available at https://www.eclipse.org/legal/epl-2.0/ 
+ * Contributors:
+ * Arthur Daussy - initial API and implementation.
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java:8]: Change "Arthur Daussy" to "Lucas Chaigne"

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+import fr.adaussy.permadeler.model.Permadeler.KnowledgeBase;
+import fr.adaussy.permadeler.model.Permadeler.PermadelerFactory;
+import fr.adaussy.permadeler.model.Permadeler.Plantation;
+import fr.adaussy.permadeler.model.Permadeler.Planting;
+import fr.adaussy.permadeler.model.Permadeler.Root;
+import fr.adaussy.permadeler.model.Permadeler.Species;
+import fr.adaussy.permadeler.model.Permadeler.Zone;
+
+/**
+ * Tests for the service {@link ValidationService#validatePlantationCompatibility(Area)}
```
Comment (adaussy) [inline bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java:35]: No this class whill hold all test on all method in ValidationService (not only validatePlantationCompatibility)

Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java]: ok, it make sense. But then, should we do also some kind of class hierarchy for unit tests ?
like : 
- public class ValidationServiceTest
     - public class ValidatePlantationCompatibilty
          - test1
          - test 2...
     - public class WhateverServiceWillComeAfter
     - private static class InputModelHelper

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java (hunk 10 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java]: done

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java (hunk 9 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java]: ok, except for nbPlantations and sFactory which are used only in the InputModelHelper class and for which it does make sense to be be static. is that right ?

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java (hunk 8 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java]: done

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java (hunk 9 above)
Comment (adaussy) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java]: Yes that's right

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java (hunk 11 above)
Comment (adaussy) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java]: What we usually do it. On test class for one service class. All test methods are public method anotated with @test. No no need for sub inner class. The name of the method could be used to identify the method under test : checkValidatePlantationCompatibiltyWith<Context>
For example : checkValidatePlantationCompatibiltyWith2CompatGenuses

File: bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java (hunk 5 above)
Comment (adaussy) [reply bundles/fr.adaussy.permadeler.model.design/src/fr/adaussy/permadeler/model/design/services/ValidationService.java]: Ok lets keep it this way for now

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java (hunk 11 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java]: ok, i changed the names so it's clearer

File: bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java (hunk 9 above)
Comment (lucas-ch) [reply bundles/fr.adaussy.permadeler.model.design.tests/src/fr/adaussy/permadeler/model/design/services/ValidationServiceTest.java]: ok

**Pull request merged**